provider/docker.go

```diff
@@ -303,6 +314,120 @@ func (provider *Docker) hasCircuitBreakerLabel(container dockerData) bool {
 	return true
 }
 
+// Regexp used to extract the name of the service and the name of the property for this service
+// All properties are under the format traefik.<servicename>.frontent.*= except the port/weight/protocol directly after traefik.<servicename>.
+var servicesPropertiesRegexp = regexp.MustCompile(`^traefik\.(?P<service_name>.*?)\.(?P<property_name>port|weight|protocol|frontend\.(.*))$`)
+
+// Map of services properties
+// we can get it with label[serviceName][propertyName] and we got the propertyValue
+type labelServiceProperties map[string]map[string]string
+
+// Check if for the given container, we find labels that are defining services
+func (provider *Docker) hasServices(container dockerData) bool {
+	return len(extractServicesLabels(container.Labels)) > 0
+}
+
+// Extract the service labels from container labels of dockerData struct
+func extractServicesLabels(labels map[string]string) labelServiceProperties {
+	v := make(labelServiceProperties)
+
+	for index, serviceProperty := range labels {
+		matches := servicesPropertiesRegexp.FindStringSubmatch(index)
+		if matches != nil {
+			result := make(map[string]string)
+			for i, name := range servicesPropertiesRegexp.SubexpNames() {
+				if i != 0 {
+					result[name] = matches[i]
+				}
+			}
+			serviceName := result["service_name"]
+			if _, ok := v[serviceName]; !ok {
+				v[serviceName] = make(map[string]string)
+			}
+			v[serviceName][result["property_name"]] = serviceProperty
+		}
+	}
+
+	return v
+}
+
+// Gets array of service names for a given container
+func (provider *Docker) getServiceNames(container dockerData) []string {
+	labelServiceProperties := extractServicesLabels(container.Labels)
+	keys := make([]string, 0, len(labelServiceProperties))
+	for k := range labelServiceProperties {
+		keys = append(keys, k)
+	}
+	return keys
+}
+
+// Extract entrypoints from labels for a given service and a given docker container
+func (provider *Docker) getServiceEntryPoints(container dockerData, serviceName string) []string {
+	if entryPoints, ok := extractServicesLabels(container.Labels)[serviceName]["frontend.entryPoints"]; ok {
```

Comment:
ok so you prefer something more generic ?
getServiceValue $container $serviceName $key ?